File: peer-web/src/peer-config/shared-secret.ts
import { decode, encode } from 'base64-arraybuffer';
import { PeerAuth } from './peer-auth';
import { PeerConfig } from './peer-config';

const subtle = crypto.subtle;
const textEncoder = new TextEncoder();


export interface SharedSecretConfig {
    secretBase64: string;
    peerConfig: PeerConfig;
}

export async function createInitiatorConfig(): Promise<SharedSecretConfig> {
    const key = await subtle.generateKey({name: 'HMAC', hash: { name: 'SHA-256' }}, true, ['sign', 'verify']);
    const keyBytes = await subtle.exportKey('raw', key);
    const peerAuth = await SharedSecretPeerAuth.create(keyBytes);

    // For shared secret keys, we use the hash of the key as the pairing ID:
    const pairingIdBytes = await subtle.digest('SHA-256', keyBytes);
    const pairingId = encode(pairingIdBytes);

    const peerConfig: PeerConfig = {
        pairingId,
        peerAuth,
        role: 'initiator',
    };

    return {
        secretBase64: encode(keyBytes),
        peerConfig,
    };
}

export async function createResponderConfig(sharedKeyBase64: string): Promise<PeerConfig> {
    const keyBytes = decode(sharedKeyBase64)

    // For shared secret keys, we use the hash of the key as the pairing ID:
    const pairingIdBytes = await subtle.digest('SHA-256', keyBytes);
    const pairingId = encode(pairingIdBytes);

    const peerAuth = await SharedSecretPeerAuth.create(keyBytes);

    return {
        pairingId,
        peerAuth,
        role: 'responder',
    }
}

export class SharedSecretPeerAuth implements PeerAuth {
    /**
     * Creates an instance of PeerAuth where each peer knows the same shared
     * secret, and signs/verifies eachothers' messages with it.
     * 
     * @param sharedSecret raw bytes of the shared signing secret.
     */
    static async create(sharedSecret: ArrayBuffer): Promise<PeerAuth> {
        const sharedKey = await subtle.importKey(
            'raw',
            sharedSecret,
            {
                name: 'HMAC',
                hash: 'SHA-256'
            },
            false,
            ['sign', 'verify']
        );
        return new SharedSecretPeerAuth(sharedKey);
    }

    private constructor(private sharedKey: CryptoKey) {}

    async signMessage(message: string): Promise<string> {
        const signature = await subtle.sign(
            'HMAC',
            this.sharedKey,
            textEncoder.encode(message)
        );
        return encode(signature);
    }

    async verifyMessage(base64Signature: string, message: string): Promise<boolean> {
        return await subtle.verify(
            'HMAC',
            this.sharedKey,
            decode(base64Signature),
            textEncoder.encode(message)
        );
    }
}

File: peer-go/peer-config/shared_secret.go
package peerconfig

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"fmt"
	"io"

	"github.com/thingify-app/thing-rtc/peer-go/pairing"
)

const KEY_LENGTH = 64

type SharedSecretConfig struct {
	SecretBase64 string
	PeerConfig   *PeerConfig
}

func CreateInitiatorConfig() (*SharedSecretConfig, error) {
	return CreateInitiatorConfigWithRand(rand.Reader)
}

func CreateInitiatorConfigWithRand(randReader io.Reader) (*SharedSecretConfig, error) {
	keyBytes := make([]byte, KEY_LENGTH)
	_, err := randReader.Read(keyBytes)
	if err != nil {
		return nil, err
	}

	secretBase64 := base64.StdEncoding.EncodeToString(keyBytes)
	peerConfig, err := createPeerConfig(secretBase64, Initiator)
	if err != nil {
		return nil, err
	}

	return &SharedSecretConfig{
		SecretBase64: secretBase64,
		PeerConfig:   peerConfig,
	}, nil
}

func CreateInitiatorConfigWithSecret(sharedKeyBase64 string) (*PeerConfig, error) {
	return createPeerConfig(sharedKeyBase64, Initiator)
}

func CreateResponderConfig(sharedKeyBase64 string) (*PeerConfig, error) {
	return createPeerConfig(sharedKeyBase64, Responder)
}

func createPeerConfig(sharedKeyBase64 string, role Role) (*PeerConfig, error) {
	keyBytes, err := base64.StdEncoding.DecodeString(sharedKeyBase64)
	if err != nil {
		return nil, err
	}

	if len(keyBytes) != KEY_LENGTH {
		return nil, fmt.Errorf("Incorrect key length (%d)!", len(keyBytes))
	}

	pairingId := calculatePairingId(keyBytes)
	peerAuth := &sharedSecretPeerAuth{
		sharedKey: keyBytes,
	}

	return &PeerConfig{
		PeerAuth:  peerAuth,
		PairingId: pairingId,
		Role:      role,
	}, nil
}

func calculatePairingId(keyBytes []byte) string {
	// For shared secret keys, we use the hash of the key as the pairing ID:
	pairingIdBytes := sha256.Sum256(keyBytes)
	return base64.StdEncoding.EncodeToString(pairingIdBytes[:])
}

type sharedSecretPeerAuth struct {
	sharedKey []byte
}

func (s *sharedSecretPeerAuth) SignMessage(message string) (string, error) {
	mac := hmac.New(sha256.New, s.sharedKey)
	mac.Write([]byte(message))
	signatureBytes := mac.Sum(nil)

	return base64.StdEncoding.EncodeToString(signatureBytes), nil
}

func (s *sharedSecretPeerAuth) VerifyMessage(base64Signature string, message string) bool {
	signatureBytes, err := base64.StdEncoding.DecodeString(base64Signature)
	if err != nil {
		return false
	}

	mac := hmac.New(sha256.New, s.sharedKey)
	mac.Write([]byte(message))
	expected := mac.Sum(nil)
	return hmac.Equal(expected, signatureBytes)
}

func (*sharedSecretPeerAuth) GenerateNonce() string {
	return pairing.GenerateNonce()
}
